### PR TITLE
feat(python): expose ObjectStoreProvider registration in pylance

### PR DESCRIPTION
**Pull request by farazshaikh**
## Summary

Exposes Lance's existing Rust `ObjectStoreProvider` registration through the `pylance` bindings, so an **out-of-tree** provider can be registered at process runtime and used transparently by `lance.dataset(...)` / `lance.write_dataset(...)`.

The motivating consumer is an on-node NVMe cache provider that lives in a **separate wheel** — that provider is *not* part of this PR. This PR is only the generic registration surface in `pylance`.

## What this adds (all under `python/`)

| Python surface | Rust | Notes |
| --- | --- | --- |
| `_ObjectStoreRegistry()` | `Arc<ObjectStoreRegistry>` | inherits built-in schemes |
| `registry.register_provider(scheme, provider)` | `ObjectStoreRegistry::insert(...)` | add/override a scheme |
| `Session(store_registry=...)` | threaded into `Session::with_cache_backends(index_cache, metadata_cache, store_registry)` | defaults to `Default::default()`, backward compatible |
| `_ObjectStoreProvider.memory()` | in-tree `MemoryStoreProvider` | fully functional |
| `_ObjectStoreProvider(py_obj)` | Python-callable bridge | **stubbed** (raises on dispatch) — follow-up |
| `_ObjectStoreProvider.from_capsule(capsule)` | adopt `Arc<dyn ObjectStoreProvider>` from another wheel | see below |

Write path: a `Session` carrying a custom registry reaches `write_dataset` via the existing `session=` parameter (the Rust `get_write_params` already extracts it), so writes to a custom scheme resolve through the provider.

Smoke test: `python/python/tests/test_custom_registry.py` (4 cases) registers a made-up `test-mem://` scheme backed by `MemoryStoreProvider` and round-trips a dataset.

## The `PyCapsule` handoff (`from_capsule`)

An out-of-tree provider is compiled Rust, so we hand its `Arc<dyn ObjectStoreProvider>` straight across the Python boundary in a [`PyCapsule`](https://docs.python.org/3/c-api/capsule.html) rather than round-tripping I/O through Python. The external wheel produces a capsule named `lance_object_store_provider`; `from_capsule` validates the capsule name before dereferencing, then clones the `Arc` (the capsule keeps its own strong ref and drops it on GC — no double-free, no leak).

### Why this is sound, and the caveat
- The capsule name is validated before the pointer is dereferenced, so a foreign/mismatched capsule raises `ValueError` instead of reading arbitrary memory.
- **ABI lockstep**: this is a Rust→Rust handoff, sound only when the producer wheel and `pylance` are built against the identical `lance-io` / `object_store` source + resolved dependency versions + toolchain (i.e. **co-compiled**). That guarantees all three — the OSP, pylance, and lance-rust — agree on the vtable layout of `dyn ObjectStoreProvider`. It is **not** the general `object_store` FFI ([apache/arrow-rs-object-store#286](https://github.com/apache/arrow-rs-object-store/issues/286)) and does not round-trip through the GIL.

This deliberately does **not** attempt truly dynamic loading of a provider from a universal shared library — that would require standardizing the ABI of `dyn ObjectStoreProvider` (and `dyn` trait objects generally). For a distributable form, arrow-rs#286 or an in-Lance prototype would be the clean long-term path — happy to discuss which you'd prefer.

## Testing

The change has been exercised for **over a week** by a large-scale ML training stack, reading a logits Lance table over Azure through the (co-compiled, out-of-tree) provider.

## Notes for reviewers
- Rebased onto current `main`; the only non-trivial merge was threading `store_registry` into the newer `Session::with_cache_backends(...)` (which already carries a registry as its third argument, previously `Default::default()`).
- The out-of-tree provider and its `_provider_capsule` producer live in a separate wheel and are intentionally excluded.
- Opened as **draft** for design feedback on the `from_capsule` / co-compilation approach vs. a distributable ABI.

Related: apache/arrow-rs-object-store#286.

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -7820,6 +7821,7 @@ def write_dataset(
         "external_blob_mode": external_blob_mode,
         "allow_external_blob_outside_bases": allow_external_blob_outside_bases,
         "blob_pack_file_size_threshold": blob_pack_file_size_threshold,
+        "session": session,
```
[the remark this answers is not in the document]
Comment (farazshaikh) [reply python/python/lance/dataset.py:7824]: Looking into it - will follow up with a fix.

[the remark this answers is not in the document]
Comment (farazshaikh) [reply python/python/lance/dataset.py:7824]: Based on the reproducer:

1. The commit-handler selection is a pre-existing condition. `commit_handler_from_url` resolves the handler from the URL scheme string: known schemes get safe handlers, and everything else falls to `UnsafeCommitHandler` ([code](https://github.com/lance-format/lance/blob/ddc2f376ae7ca2208670b028cf0bb202bc5761c7/rust/lance-table/src/io/commit.rs#L1091-L1181)). This PR only enables registering a new URL-scheme-based provider, and such a scheme behaves exactly like any non-allowlisted scheme does today.

2. Tachyon is the only out-of-tree provider for now, and it resolving to `UnsafeCommitHandler` is fine because the Tachyon provider does not support writes: it is a read-only cache whose store returns `NotImplemented` for `put_opts`, `put_multipart_opts`, `delete_stream`, and `copy_opts`. A commit through it fails at the put layer, before any commit-handler logic runs.

3. In the long run, dynamically registered providers should be able to specify their commit handler from a set of preconfigured ones (for example, a conditional-put-capable store selecting `ConditionalPutCommitHandler`), with `UnsafeCommitHandler` remaining the default. That is an additive follow-up, tracked separately from this registration PR.

**Review by westonpace — APPROVED** [review]
This is a fine initial starting point.  It's a baby step towards eventual object registry support.  It sounds like we still need:

* Allow for object stores to specify what commit handler to use (if nothing else we should change the default for unrecognized schemes to the put-if-not-exists handler).
* A stable ABI for object stores
* More public APIs.  We shouldn't have to underscore-prefixed methods like `_ObjectStoreRegistry` and `_Session`.  We should either expose these more publicly or add factory methods.  Still, this is probably best to tackle last, one the above two points are finished.

As a starting point this will allow for experimentation and working through any issues we might find on the read paths.
